File: Pod/Classes/FNKYAxis.m
//
//  FNKYAxis.m
//  FNKGraphs
//
//  Created by Phillip Connaughton on 11/26/14.
//  Copyright (c) 2014 fnk. All rights reserved.
//

#import "FNKYAxis.h"

@implementation FNKYAxis

-(void) drawAxis:(UIView*) view
{
    UIBezierPath* bezPath = [[UIBezierPath alloc] init];
    [bezPath moveToPoint:CGPointMake(self.marginLeft, 0)];
    [bezPath addLineToPoint:CGPointMake(self.marginLeft, self.graphHeight)];
    [bezPath closePath];
    
    CAShapeLayer* layer = [[CAShapeLayer alloc] init];
    layer.path = bezPath.CGPath;
    layer.fillColor = self.fillColor.CGColor;
    layer.strokeColor = self.strokeColor.CGColor;
    
    [view.layer addSublayer:layer];
    
    CABasicAnimation* pathAnimation = [CABasicAnimation animationWithKeyPath:@"strokeEnd"];
    pathAnimation.duration = 2;
    pathAnimation.fromValue = @(0);
    pathAnimation.toValue = @(1);
    
    [layer addAnimation:pathAnimation forKey:@"strokeEnd"];
    
    //Draw all lines
    CGFloat tickInterval = self.graphHeight / self.ticks;
    for (int index = 0 ; index < self.ticks+1 ; index++)
    {
        UIBezierPath* bezPath = [[UIBezierPath alloc] init];
        
        CGFloat xVal = self.marginLeft;
        CGFloat yVal = (index * tickInterval);
        
        [bezPath moveToPoint:CGPointMake(xVal, yVal)];
        [bezPath addLineToPoint:CGPointMake(xVal + self.graphWidth, yVal)];
        
        CAShapeLayer* layer  = [[CAShapeLayer alloc] init];
        layer.path = bezPath.CGPath;
        layer.fillColor = self.tickFillColor.CGColor;
        layer.strokeColor = self.tickStrokeColor.CGColor;
        
        [view.layer addSublayer:layer];
        
        CABasicAnimation* pathAnimation = [CABasicAnimation animationWithKeyPath:@"strokeEnd"];
        pathAnimation.duration = self.animationDuration;
        pathAnimation.fromValue = @(0);
        pathAnimation.toValue = @(1);
        
        [layer addAnimation:pathAnimation forKey:@"strokeEnd"];
    }
}

-(UIView*) addTicksToView:(UIView*) view
{
    CGFloat tickInterval = self.graphHeight / self.ticks;
    
    UIView* labelView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 40,  view.frame.size.height)];
    [labelView setAlpha:0.0];
    [view addSubview:labelView];
    labelView.translatesAutoresizingMaskIntoConstraints = NO;
    NSDictionary *labelViewDictionary = NSDictionaryOfVariableBindings(labelView);
    
    NSArray *widthConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"|-0-[labelView(40)]"
                                                                        options:0
                                                                        metrics:nil
                                                                          views:labelViewDictionary];
    [view addConstraints:widthConstraints];
    
    NSArray *heightConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:|-0-[labelView]-0-|"
                                                                         options:0
                                                                         metrics:nil
                                                                           views:labelViewDictionary];
    [view addConstraints:heightConstraints];
    
    [view layoutSubviews];
    
    for (int index = 0 ; index < self.ticks ; index++)
    {
        if (index == self.ticks)
        {
            continue;
        }
        
        CGFloat yLoc = (index * tickInterval);
        CGFloat yVal = (index * tickInterval);
        
        //Okay those are the ticks. Now we need the labels
        UILabel* tickLabel = [[UILabel alloc] init];
        
        CGFloat originalVal = ((self.graphHeight - yVal) / self.scaleFactor) + self.axisMin;
        tickLabel.text = self.tickFormat(originalVal);
        tickLabel.textAlignment = NSTextAlignmentCenter;
        tickLabel.font = self.tickFont;
        [tickLabel setTextColor: self.tickLabelColor ? self.tickLabelColor : [UIColor blackColor]];
        [tickLabel setAdjustsFontSizeToFitWidth:YES];
        [tickLabel setMinimumScaleFactor:.5];
        [tickLabel setBackgroundColor:[[UIColor whiteColor] colorWithAlphaComponent:.5]];
        [labelView addSubview:tickLabel];
        
        tickLabel.translatesAutoresizingMaskIntoConstraints = NO;
        NSDictionary *viewDictionary = NSDictionaryOfVariableBindings(tickLabel);
        
        NSArray *widthConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"|-2-[tickLabel(36)]-2-|"
                                                                            options:0
                                                                            metrics:nil
                                                                              views:viewDictionary];
        [labelView addConstraints:widthConstraints];
        
        //TODO: This is a bit of a hack since I don't know what the height of the label is actually going to be
        NSArray *heightConstraints = [NSLayoutConstraint constraintsWithVisualFormat:[NSString stringWithFormat:@"V:|-(%f)-[tickLabel]", yLoc - 5]
                                                                             options:0
                                                                             metrics:nil
                                                                               views:viewDictionary];
        [labelView addConstraints:heightConstraints];
        
        [labelView layoutSubviews];
    }
    
    [UIView animateWithDuration:1
                     animations:^{
                         [labelView setAlpha:1.0];
                         
                     }];
    
    
    
    return labelView;
}

@end

File: Pod/Classes/FNKYAxis.swift
//
//  FNKYAxis.swift
//  Pods
//
//  Created by Phillip Connaughton on 11/21/15.
//
//

import Foundation

class FNKYAxis: FNKAxis
{
    let tickType: FNKYAxisTickType
    let paddingPercentage: Float
    
    init(marginLeft: CGFloat, marginBottom: CGFloat, scaleFactor: CGFloat, axisMin: CGFloat, tickType: FNKYAxisTickType,paddingPercentage: Float, graphDimensions: CGRect)
    {
        self.tickType = tickType
        self.paddingPercentage = paddingPercentage
        super.init(marginLeft: marginLeft, marginBottom: marginBottom, scaleFactor: scaleFactor, axisMin: axisMin, graphDimensions: graphDimensions)
    }
    
    override func drawAxis(view: UIView) {
        
        let bezPath = UIBezierPath()
        bezPath.moveToPoint(CGPointMake(self.graphDimensions.width, 0))
        bezPath.addLineToPoint(CGPointMake(self.marginLeft, self.graphDimensions.height))
        bezPath.closePath()
        
        let layer = CAShapeLayer()
        layer.path = bezPath.CGPath
        layer.fillColor = self.fillColor.CGColor
        layer.strokeColor = self.strokeColor.CGColor
        
        view.layer.addSublayer(layer)
        
        let pathAnimation = CABasicAnimation(keyPath: "strokeEnd")
        pathAnimation.duration = 2
        pathAnimation.fromValue = 0.0
        pathAnimation.toValue = 1.0
        
        layer.addAnimation(pathAnimation, forKey: "strokeEnd")
        
        //Draw all lines
        let tickInterval = self.graphDimensions.height / CGFloat(self.ticks)
        for index in 0...self.ticks+1
        {
            let bezPath = UIBezierPath()
            
            let xVal = self.marginLeft
            let yVal = CGFloat(index) * tickInterval
            
            bezPath.moveToPoint(CGPointMake(xVal, yVal))
            
            if(self.tickType == .Outside)
            {
                bezPath.addLineToPoint(CGPointMake(self.graphDimensions.width, yVal))
            }
            else if(self.tickType == .Behind)
            {
                bezPath.addLineToPoint(CGPointMake(xVal + self.graphDimensions.width, yVal))
            }
            else if(self.tickType == .Above)
            {
                bezPath.addLineToPoint(CGPointMake(xVal + self.graphDimensions.width, yVal))
            }
            
            let layer  = CAShapeLayer()
            layer.path = bezPath.CGPath
            layer.fillColor = self.tickFillColor.CGColor
            layer.strokeColor = self.tickStrokeColor.CGColor
            
            view.layer.addSublayer(layer)
            
            let pathAnimation = CABasicAnimation(keyPath: "strokeEnd")
            pathAnimation.duration = self.animationDuration
            pathAnimation.fromValue = 0.0
            pathAnimation.toValue = 1.0
            
            layer.addAnimation(pathAnimation, forKey:"strokeEnd")
        }
    }
    
    override func addTicksToView(view: UIView, tickFormat: (CGFloat) -> (String)) -> UIView?
    {
        let tickInterval = self.graphDimensions.height / CGFloat(self.ticks)
        
        let labelView = UIView(frame:CGRectMake(0, 0, 40,  view.frame.size.height))
        labelView.alpha = 0.0
        view.addSubview(labelView)
        labelView.translatesAutoresizingMaskIntoConstraints = false
        let labelViewMap = ["labelView" :labelView]
        
        if(self.tickType == .Behind)
        {
            let widthConstraints = NSLayoutConstraint.constraintsWithVisualFormat(String(format: "|-0-[labelView(%ld)]", arguments: [self.marginLeft]), options: .AlignmentMask, metrics: nil, views: labelViewMap)
                
            view.addConstraints(widthConstraints)
        }
        else if(self.tickType == .Above)
        {
            let widthConstraints = NSLayoutConstraint.constraintsWithVisualFormat("|-0-[labelView]",
            options:.AlignmentMask,
            metrics:nil,
            views:labelViewMap)
            view.addConstraints(widthConstraints)
        }
        
        let heightConstraints = NSLayoutConstraint.constraintsWithVisualFormat("V:|-0-[labelView]-0-|", options:.AlignmentMask, metrics:nil, views:labelViewMap)
        view.addConstraints(heightConstraints)
        
        view.layoutSubviews()
        
        for index in 0...self.ticks
        {
            if (index == self.ticks)
            {
                continue
            }
        
            let yLoc = self.tickType == .Above ? (CGFloat(index) * tickInterval) - self.tickFont.capHeight :  (CGFloat(index) * tickInterval);
            let yVal = self.tickType == .Above ? (CGFloat(index) * tickInterval) :  (CGFloat(index) * tickInterval);
            
            //Okay those are the ticks. Now we need the labels
            let tickLabel = UILabel()
            
            let originalVal = ((self.graphDimensions.height - yVal) / self.scaleFactor) + self.axisMin
            tickLabel.text = tickFormat(originalVal) 
            
            if(self.tickType == .Above)
            {
                tickLabel.textAlignment = .Left;
            }
            else
            {
                tickLabel.textAlignment = .Right;
            }
        
            tickLabel.font = self.tickFont
            tickLabel.textColor = self.tickLabelColor
            tickLabel.adjustsFontSizeToFitWidth = true
            tickLabel.minimumScaleFactor = 0.5
            tickLabel.backgroundColor = UIColor.whiteColor().colorWithAlphaComponent(0.5)
            labelView.addSubview(tickLabel)
            
            tickLabel.translatesAutoresizingMaskIntoConstraints = false
            let labelViewMap = ["labelView" :labelView]
            
            var constraintString: String
            if(self.tickType == .Above)
            {
                constraintString = String("|-(%f)-[tickLabel]-2-|", self.marginLeft)
            }
            else
            {
                constraintString = "|-0-[tickLabel]-2-|"
            }
        
            let widthConstraints = NSLayoutConstraint.constraintsWithVisualFormat(constraintString, options:.AlignmentMask, metrics:nil, views:labelViewMap)
            labelView.addConstraints(widthConstraints)
            
            //TODO: This is a bit of a hack since I don't know what the height of the label is actually going to be
            constraintString = String("V:|-(%f)-[tickLabel]", yLoc - 5)
            let heightConstraints = NSLayoutConstraint.constraintsWithVisualFormat(constraintString, options:.AlignmentMask, metrics:nil, views:labelViewMap)
            labelView.addConstraints(heightConstraints)
            
            labelView.layoutSubviews()
        }

        UIView.animateWithDuration(1.0) { () -> Void in
            labelView.alpha = 1.0
        }
        
        return labelView;
    }
}